chromium_src/base/debug/alias.h

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+#define BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+
+#include "base/containers/span.h"
+#include "base/memory/raw_ptr_exclusion.h"
+#include "base/memory/stack_allocated.h"
+#include "base/ranges/algorithm.h"
+
+#include "src/base/debug/alias.h"  // IWYU pragma: export
+
+namespace base::debug {
+
+// StackObjectCopy creates a byte-for-byte copy of an object on the stack,
+// allowing the object to be inspected in crash dumps even if the original
+// object is optimized away by the compiler or is stored in protected memory.
+// The copy maintains proper alignment and can be viewed as the original type in
+// a debugger.
+template <typename T>
+class StackObjectCopy {
+  STACK_ALLOCATED();
+
+ public:
+  explicit StackObjectCopy(const T* original) {
+    if (original) {
+      // SAFETY: `original` is a valid pointer to a T object, and `kSize` is the
+      // size of T.
+      auto original_object_memory = UNSAFE_BUFFERS(
+          base::make_span(reinterpret_cast<const uint8_t*>(original), kSize));
```

Comment:
This is a valid case where `reinterpret_cast` is the right way of doing things. I'm working on memory directly to copy it on stack so it can be visible in the memory dump. I'm not doing a C++ object copy here to pass it somewhere else and/or call some methods on it.

Reply:
I get what you're saying, but the issue is that the copy itself is undefined behaviour as per the standard, and then the reinterpret cast becomes undefined behaviour by consequence. C++ is really annoying about this, and limits a lot what one can do with this type of thing.

Reply:
> I get what you're saying, but the issue is that the copy itself is undefined behaviour as per the standard, and then the reinterpret cast becomes undefined behaviour by consequence. C++ is really annoying about this, and limits a lot what one can do with this type of thing.

Please tell me at what exact line we get undefined behavior here? I'm copying an array of bytes which is a valid and defined operation. I understand your concern that some optimization may do something wrong here, but we're operating at byte level, so I don't see a single place where things can go wrong.

Reply:
The way I see it this use case is a legal one.
`uint8_t` is similar to `unsigned char`. It's one of the types that is accessible via *reinterpret_cast:
https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_accessibility

Also chromium code has multiple examples of similar access to an object byte representation using the same approach:
https://chromium.googlesource.com/chromium/src/+/refs/heads/main/base/feature_list.cc#169